File: scala/p31/p31.scala
object P31 {

  val coins = Array(1, 2, 5, 10, 20, 50, 100, 200);

  def main(args: Array[String]) = {
    val amount = args(0).toInt
    println(s"Total number of hands: ${ways(amount, 7)}")
  }


  def ways(t: Int, i: Int):Int = {
    if (i <= 0) {
      return 1
    }

    var r = 0
    var tt = t
    while(tt >= 0){
      r = r + ways(tt, i - 1)
      tt = tt - coins(i)
    }
    return r
  }
}

File: go/src/P31/P31.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func main() {
	if len(os.Args) == 1 {
		fmt.Printf("Please enter amount\n")
		return
	}

	// for i := 0; i < len(os.Args); i++ {
	//	fmt.Printf("%d: %s\n", i, os.Args[i])
	// }

	amount, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println(err)
		os.Exit(2)
	}

	fmt.Printf("result is %d\n", ways(amount, 7))
	return
}

var coins = [8]int{1, 2, 5, 10, 20, 50, 100, 200}

// t is target amount, i is current index in coins array(start from end)
func ways(t int, i int) int {
	if i <= 0 {
		return 1
	}

	r := 0
	for t >= 0 {
		r = r + ways(t, i-1)
		t = t - coins[i]
	}
	return r
}
